flamingpy/noise/cv.py

```diff
@@ -0,0 +1,649 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Continuous-variable noise model classes."""
+
+# pylint: disable=import-outside-toplevel
+
+import numpy as np
+from numpy.random import default_rng
+import scipy.sparse as sp
+
+from flamingpy.cv.ops import invert_permutation, SCZ_mat, SCZ_apply, splitter_symp
+from flamingpy.cv.gkp import GKP_binner, Z_err_cond
+
+# pylint: disable=too-many-instance-attributes
+class CVLayer:
+    """A class for applying to a code (or an EGraph) a physical layer of
+    continuous-variable noise.
+
+    Associates the nodes of an EGraph with continuous-variable quantum states,
+    and its edges with continuous-variable CZ gates.
+
+    For now, only a hybrid state of p-squeezed and GKP states is considered.
+
+    Args:
+        code (SurfaceCode or EGraph): the code object (so that code.graph is
+            an EGraph) or an EGraph directly.
+        state (dict, optional): the dictionary of all non-GKP states and their
+            indices, of the form {'state': []}. By default, all states are
+            GKP states.
+        p_swap (float, optional): if supplied, the probability of a node being
+            a p-squeezed state. Overrides the indices given in state.
+        rng (numpy.random.Generator, optional): a random number generator
+            following the NumPy API. It can be seeded for reproducibility.
+            By default, numpy.random.default_rng is used without a fixed seed.
+
+    Attributes:
+        egraph (EGraph): the unerlying graph representation.
+        _N (int): the number of qubits in the lattice.
+        _adj (sp.sparse.csr_matrix): the adjacency matrix of egraph.
+        _states (dict): states along with their indices.
+        _delta (float): the delta from the Args above (after noise applied)
+        _adj (array): adjancency matrix of the underlying graph.
+        to_points (dict): pointer to self.egraph.to_points, the dictionary from
+            indices to coordinates.
+    """
+
+    def __init__(self, code, states=None, p_swap=0, rng=default_rng()):
+        """Initialize the CVGraph."""
+        if code.__class__.__name__ == "EGraph":
+            self.egraph = code
+        else:
+            self.egraph = code.graph
```

Comment:
Motivation for this was that the noise layer could be applied to the code (modifying code.graph) or to a graph independently of the qubit code (like the macronized version of the graph).

I did not use `isinstance` because:
- It would require me to import an EGraph (and other classes) into the file 
- It is potentially slower?

Reply:
more imports and the call to `isinstance` are not a big issue in term of performance. I would favour the clearer approach of using `isinstance`.

Reading your comment, I know think that the argument should always be the graph directly then. The less possibilities, the simpler the usage.

Reply:
@soosub @nariman87 wanted to get your thoughts on this. In this case the only reason we would be importing EGraph into the file is for this type comparison. Would this still be preferred, or would my solution (`if code.__class__.__name__ == "EGraph":`) be ok?

Reply:
If importing the `EGraph` is not an issue I would go for the `isinstance(code, EGraph)`. There are several reasons:

- This is comparing the name of the classes: if a new class inherits from `EGraph`, i.e. `NewEGraph`, the check will fail when it should succeed.
- A crazy dev 🤪 decides to overwrite the `__class__` method of `EGraph` and return an int. The check will pass but you'll be dealing with a completely different object.
- Crazy dev number two 🙃 decides to create a new class named `EGraph` with a completely different functionality. Check will pass because the classes have the same name but in fact the objects are completely different.
- PEP8 states 📖  _"Object type comparisons should always use `isinstance()` instead of comparing types directly"_. I guess this recommendation also outpoints comparing classes by names. 

If importing `EGraph` is an issue because it creates circular dependencies, then the import statement can be deferred and place right before the check.

Reply:
Sorry, folks, just saw this discussion. I second @sduquemesa views here. BTW, I might be that Crazy Dev ins @sduquemesa's stories !!! :-)

Reply:
Thanks for the great answer @sduquemesa ! I'm convinced.